crates/stages/stages/src/stages/era.rs

```diff
@@ -0,0 +1,603 @@
+use crate::{stages, StageCheckpoint, StageId};
+use alloy_primitives::{BlockHash, BlockNumber};
+use futures_util::{Stream, StreamExt};
+use reqwest::{Client, Url};
+use reth_config::config::EtlConfig;
+use reth_db_api::{table::Value, tables, transaction::DbTxMut, DbTxUnwindExt};
+use reth_era::era1_file::Era1Reader;
+use reth_era_downloader::{read_dir, EraClient, EraMeta, EraStream, EraStreamConfig};
+use reth_era_utils as era;
+use reth_etl::Collector;
+use reth_primitives_traits::{FullBlockBody, FullBlockHeader, NodePrimitives};
+use reth_provider::{
+    BlockHashReader, BlockReader, BlockWriter, DBProvider, HeaderProvider,
+    StaticFileProviderFactory, StaticFileWriter, StorageLocation,
+};
+use reth_stages_api::{ExecInput, ExecOutput, Stage, StageError, UnwindInput, UnwindOutput};
+use reth_static_file_types::StaticFileSegment;
+use reth_storage_errors::ProviderError;
+use std::{
+    fmt::{Debug, Formatter},
+    path::PathBuf,
+    task::{ready, Context, Poll},
+};
+
+type Item<BH, BB> = Box<dyn Iterator<Item = eyre::Result<(BH, BB)>> + Send + Sync + Unpin>;
+type ThreadSafeEraStream<BH, BB> =
+    Box<dyn Stream<Item = eyre::Result<Item<BH, BB>>> + Send + Sync + Unpin>;
+
+/// The [ERA1](https://github.com/eth-clients/e2store-format-specs/blob/main/formats/era1.md)
+/// pre-merge history stage.
+///
+/// Imports block headers and bodies from genesis up to the last pre-merge block. Receipts are
+/// generated by execution. Execution is not done in this stage.
+pub struct EraStage<BH, BB, F> {
+    source: Option<F>,
+    hash_collector: Collector<BlockHash, BlockNumber>,
+    item: Option<Item<BH, BB>>,
+    stream: Option<ThreadSafeEraStream<BH, BB>>,
+    last_block_height: Option<BlockNumber>,
+}
+
+trait EraStreamFactory<BH, BB> {
+    fn create(self, input: ExecInput) -> Result<ThreadSafeEraStream<BH, BB>, StageError>;
+    fn cleanup(&self) {}
+}
+
+impl<BH, BB> EraStreamFactory<BH, BB> for EraImportSource
+where
+    BH: FullBlockHeader + Value,
+    BB: FullBlockBody<OmmerHeader = BH>,
+{
+    fn create(self, input: ExecInput) -> Result<ThreadSafeEraStream<BH, BB>, StageError> {
+        Ok(match self {
```

Comment:
can we make this in two steps?
i'd like to avoid an `ok(` that spans 50 lines

Reply:
Addressed in 08442949